### ISSUE TITLE
[Bug]: `StaleDataError` when loading, modifying and then saving an object

### ISSUE THREAD

**Issue by Quarub** [Mause/duckdb_engine#311]
### What happened?

When saving a modified object that was created in a previous `Session` I get a `StaleDataError` i.e. can not `UPDATE` the object in the DB.

Everything works when running the query generated by SQAlchemy manually in SQAlchemy, in the REPL of `duckdb` and when running the code with `sqlite` instead. 

A small test case:
```python
import sqlalchemy
from sqlalchemy import Column, Integer, String, select
from sqlalchemy.orm import Mapped, Session, declarative_base

Base = declarative_base()
class CloseTestTable(Base):
    __tablename__ = "close_test_table"

    id = Column(Integer(), 
            sqlalchemy.Sequence("id_seq"),
            primary_key=True)
    astring: Mapped[str] = Column(String(1))

duck = "duckdb:///duck.db"
sqlite = "sqlite:///sqlite.db"
engine = sqlalchemy.create_engine(duck, echo=True, future=True)
Base.metadata.create_all(engine)

with Session(engine) as session:
    c = CloseTestTable()
    session.add(c)
    session.commit()
    session.close()

with Session(engine) as session:
    if c := session.get(CloseTestTable, 1):
        c.astring = "a"
        session.flush()
        session.commit()
    # Same result when getting the Object via select
    #c = session.scalar(select(CloseTestTable).where(CloseTestTable.id == 1))
    
    # Works with the textual version
    #session.execute(sqlalchemy.text("UPDATE close_test_table SET astring='a' WHERE close_test_table.id = 1")).fetchone()
    #session.commit()
```
Output duckdb:
```
2022-07-01 17:11:11,012 INFO sqlalchemy.engine.Engine select current_schema()
2022-07-01 17:11:11,012 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-07-01 17:11:11,013 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2022-07-01 17:11:11,013 INFO sqlalchemy.engine.Engine select relname from pg_class c join pg_namespace n on n.oid=c.relnamespace where pg_catalog.pg_table_is_visible(c.oid) and relname=?
2022-07-01 17:11:11,013 INFO sqlalchemy.engine.Engine [dialect duckdb+psycopg2 does not support caching 0.00006s] ('close_test_table',)
2022-07-01 17:11:11,018 INFO sqlalchemy.engine.Engine SELECT relname FROM pg_class c join pg_namespace n on n.oid=c.relnamespace where relkind='S' and n.nspname=? and relname=?
2022-07-01 17:11:11,018 INFO sqlalchemy.engine.Engine [dialect duckdb+psycopg2 does not support caching 0.00006s] ('main', 'id_seq')
2022-07-01 17:11:11,020 INFO sqlalchemy.engine.Engine CREATE SEQUENCE id_seq START WITH 1
2022-07-01 17:11:11,020 INFO sqlalchemy.engine.Engine [no key 0.00004s] ()
2022-07-01 17:11:11,020 INFO sqlalchemy.engine.Engine 
CREATE TABLE close_test_table (
	id INTEGER NOT NULL, 
	astring VARCHAR(1), 
	PRIMARY KEY (id)
)


2022-07-01 17:11:11,020 INFO sqlalchemy.engine.Engine [no key 0.00004s] ()
2022-07-01 17:11:11,021 INFO sqlalchemy.engine.Engine COMMIT
2022-07-01 17:11:11,021 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2022-07-01 17:11:11,022 INFO sqlalchemy.engine.Engine select nextval('id_seq')
2022-07-01 17:11:11,022 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-07-01 17:11:11,022 INFO sqlalchemy.engine.Engine INSERT INTO close_test_table (id, astring) VALUES (?, ?)
2022-07-01 17:11:11,022 INFO sqlalchemy.engine.Engine [dialect duckdb+psycopg2 does not support caching 0.00023s] (1, None)
2022-07-01 17:11:11,022 INFO sqlalchemy.engine.Engine COMMIT
2022-07-01 17:11:11,023 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2022-07-01 17:11:11,023 INFO sqlalchemy.engine.Engine SELECT close_test_table.id AS close_test_table_id, close_test_table.astring AS close_test_table_astring 
FROM close_test_table 
WHERE close_test_table.id = ?
2022-07-01 17:11:11,023 INFO sqlalchemy.engine.Engine [dialect duckdb+psycopg2 does not support caching 0.00005s] (1,)
2022-07-01 17:11:11,024 INFO sqlalchemy.engine.Engine UPDATE close_test_table SET astring=? WHERE close_test_table.id = ?
2022-07-01 17:11:11,024 INFO sqlalchemy.engine.Engine [dialect duckdb+psycopg2 does not support caching 0.00005s] ('a', 1)
2022-07-01 17:11:11,024 INFO sqlalchemy.engine.Engine ROLLBACK
Traceback (most recent call last):
  File "/tmp/example.py", line 30, in <module>
    session.flush()
  File "/tmp/venv/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 3383, in flush
    self._flush(objects)
  File "/tmp/venv/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 3522, in _flush
    with util.safe_reraise():
  File "/tmp/venv/lib/python3.10/site-packages/sqlalchemy/util/langhelpers.py", line 70, in __exit__
    compat.raise_(
  File "/tmp/venv/lib/python3.10/site-packages/sqlalchemy/util/compat.py", line 208, in raise_
    raise exception
  File "/tmp/venv/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 3483, in _flush
    flush_context.execute()
  File "/tmp/venv/lib/python3.10/site-packages/sqlalchemy/orm/unitofwork.py", line 456, in execute
    rec.execute(self)
  File "/tmp/venv/lib/python3.10/site-packages/sqlalchemy/orm/unitofwork.py", line 630, in execute
    util.preloaded.orm_persistence.save_obj(
  File "/tmp/venv/lib/python3.10/site-packages/sqlalchemy/orm/persistence.py", line 237, in save_obj
    _emit_update_statements(
  File "/tmp/venv/lib/python3.10/site-packages/sqlalchemy/orm/persistence.py", line 1035, in _emit_update_statements
    raise orm_exc.StaleDataError(
sqlalchemy.orm.exc.StaleDataError: UPDATE statement on table 'close_test_table' expected to update 1 row(s); -1 were matched.
```

I also ran the code against `master` with the same result.
Do you have an idea of what could be the issue?

------ 
Additional information:
Environment:
```
# python3 --version
Python 3.10.5
# pip freeze
duckdb==0.4.0
duckdb-engine==0.1.11
greenlet==1.1.2
mypy==0.961
mypy-extensions==0.4.3
numpy==1.23.0
SQLAlchemy==1.4.39
sqlalchemy2-stubs==0.0.2a24
tomli==2.0.1
typing_extensions==4.2.0
```

Output for the same code for SQLite:
```
2022-07-01 16:50:17,804 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2022-07-01 16:50:17,804 INFO sqlalchemy.engine.Engine PRAGMA main.table_info("close_test_table")
2022-07-01 16:50:17,804 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-07-01 16:50:17,804 INFO sqlalchemy.engine.Engine PRAGMA temp.table_info("close_test_table")
2022-07-01 16:50:17,804 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-07-01 16:50:17,804 INFO sqlalchemy.engine.Engine 
CREATE TABLE close_test_table (
	id INTEGER NOT NULL, 
	astring VARCHAR(1), 
	PRIMARY KEY (id)
)


2022-07-01 16:50:17,804 INFO sqlalchemy.engine.Engine [no key 0.00004s] ()
2022-07-01 16:50:17,804 INFO sqlalchemy.engine.Engine COMMIT
2022-07-01 16:50:17,805 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2022-07-01 16:50:17,805 INFO sqlalchemy.engine.Engine INSERT INTO close_test_table (astring) VALUES (?)
2022-07-01 16:50:17,806 INFO sqlalchemy.engine.Engine [generated in 0.00006s] (None,)
2022-07-01 16:50:17,806 INFO sqlalchemy.engine.Engine COMMIT
2022-07-01 16:50:17,806 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2022-07-01 16:50:17,807 INFO sqlalchemy.engine.Engine SELECT close_test_table.id AS close_test_table_id, close_test_table.astring AS close_test_table_astring 
FROM close_test_table 
WHERE close_test_table.id = ?
2022-07-01 16:50:17,807 INFO sqlalchemy.engine.Engine [generated in 0.00006s] (1,)
2022-07-01 16:50:17,809 INFO sqlalchemy.engine.Engine UPDATE close_test_table SET astring=? WHERE close_test_table.id = ?
2022-07-01 16:50:17,809 INFO sqlalchemy.engine.Engine [generated in 0.00006s] ('a', 1)
2022-07-01 16:50:17,809 INFO sqlalchemy.engine.Engine COMMIT
```

### DuckDB Engine Version

0.1.11

### DuckDB Version

v0.4.0

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by Mause** [comment]
Hi, thank you so much for such a detailed issue! It looks like the root cause here is pretty simple, the duckdb api doesn't support the rowcount dbapi extension. I'm surprised this hasn't come up before, as it looks like it would break any updates to models using duckdb-engine.

For the short term, I'll disable sqlalchemy's check for this field, as that seems to resolve the issue. I'll let you know once I've put out a release with that fix.

Longer term, I'll look at seeing how difficult it is to add support for the rowcount extension in duckdb's python client.

**Comment by Mause** [comment]
I've updated master with the short term fix, which you can install with `pip install git+https://github.com/Mause/duckdb_engine#egg=duckdb-engine`. Can you please test it to ensure it fixes your use cases?

**Comment by Quarub** [comment]
@Mause It fixes my use case :+1: 
Thank you for the fast response!

**Comment by Mause** [comment]
This has now been released in v0.2.0, please feel free to let me know if you have any further issues 😁
